subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -394,6 +511,11 @@ public ProviderInternal<T> getChangingValue() throws IllegalStateException {
             throw new IllegalStateException();
         }
 
+        @Nullable
+        public SideEffect<? super T> getSideEffect() throws IllegalStateException {
```

Comment:
I am not sure that this interface is the right place to put `getSideEffect()`. I wonder if having a hierarchy would help here. In the end, this is mostly duplicating the `Value` hierarchy. So only having this method on the fixed value might make sense?

Reply:
For the alternative implementation via a hierarchy see here: https://github.com/gradle/gradle/pull/21512